### PR TITLE
Smtp mime base64/v7

### PR DESCRIPTION
**Pull request by victorjulien**
#7473 with additional fixes and cleanups after running large pcap collections with lots of odd smtp traffic.

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on report_failure.

ERROR: QA failed on arg_strap.
  

Pipeline 7688

File: src/app-layer-smtp.c [hunk 1]
```diff
@@ -1277,43 +1278,44 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
...
-    while (state->input_len > 0 && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE)) {
-        uint8_t delim_len = 0;
-        uint32_t consumed_line = 0;
-        if (lf_idx == NULL) {
-            if ((state->input_len == 1 && state->input[state->consumed] == '-') ||
-                    (state->input_len > 1 && state->input[state->consumed] == '-' &&
+    bool line_complete = false;
+    int32_t input_len = state->input_len;
+    for (int32_t i = 0;
+            i < input_len && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE); i++) {
```
Comment (victorjulien) [inline src/app-layer-smtp.c:1284]: can this be unset in this loop? Perhaps in the `SMTPProcessRequest` call? If so, lets just check for it after that call then, and break the loop?

File: src/app-layer-smtp.c [hunk 2]
```diff
@@ -1277,43 +1278,44 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
...
-            consumed_line = lf_idx - state->current_line + 1;
-            if (orig_input_len >= idx && idx > 0 && state->input[idx - 1] == 0x0d) {
-                delim_len = 2;
-            } else if (orig_input_len == 1 ||
-                       (orig_input_len >= idx && idx > 0 && state->input[idx - 1] != 0x0d)) {
-                delim_len = 1;
+            state->current_line_len = current_line_consumed - state->current_line_delimiter_len;
+            state->consumed = total_consumed;
+            state->input_len -= current_line_consumed;
+            BUG_ON(state->consumed + state->input_len != state->orig_input_len);
```
Comment (victorjulien) [inline src/app-layer-smtp.c:1312]: DEBUG_VALIDATE_BUG_ON

File: src/util-decode-mime.c [hunk 3]
```diff
@@ -1300,100 +1319,101 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
         MimeDecParseState *state)
 {
     int ret = MIME_DEC_OK;
-    uint8_t rem1 = 0, rem2 = 0;
-    uint32_t numDecoded, remaining, offset, avail, tobuf;
+    uint32_t numDecoded, remaining = len, offset = 0;
 
-    /* Track long line */
+    /* Track long line TODO should we count space padding too? */
```
Comment (victorjulien) [inline src/util-decode-mime.c:1324]: @inashivb any thought on this?

File: src/util-decode-mime.c [hunk 4]
```diff
@@ -1300,100 +1319,101 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
...
-                PrintChars(SC_LOG_DEBUG, "Base64 failed string", buf + offset, tobuf);
+                offset++;
             }
-
-            /* Update counts */
-            remaining -= tobuf;
-            offset += tobuf;
-        }
+            return MIME_DEC_OK;
+            /* If remaining is 4 by this time, it's likely due to error/spaces during processing */
```
Comment (victorjulien) [inline src/util-decode-mime.c:1406]: comment talks about error but its no longer an error that leads us here I think, just incomplete data or data with spaces

File: src/app-layer-smtp.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-smtp.c:1284]: thinking about
```diff
diff --git a/src/app-layer-smtp.c b/src/app-layer-smtp.c
index fe441c778..c19398ad8 100644
--- a/src/app-layer-smtp.c
+++ b/src/app-layer-smtp.c
@@ -1278,10 +1278,11 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
 
 static int SMTPPreProcessCommands(SMTPState *state, Flow *f, AppLayerParserState *pstate)
 {
+    DEBUG_VALIDATE_BUG_ON((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0);
+
     bool line_complete = false;
     int32_t input_len = state->input_len;
-    for (int32_t i = 0;
-            i < input_len && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE); i++) {
+    for (int32_t i = 0; i < input_len; i++) {
         if (state->input[i] == 0x0d) {
             if (i < input_len - 1 && state->input[i + 1] == 0x0a) {
                 i++;
@@ -1315,6 +1316,9 @@ static int SMTPPreProcessCommands(SMTPState *state, Flow *f, AppLayerParserState
             }
             line_complete = false;
             state->current_line_delimiter_len = 0;
+
+            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0)
+                break;
         }
     }
     return 0;
```
which I find more readable and easier to understand

**Comment by victorjulien** [comment]
Replaced by #7482

**Pull request closed without merging**